src/OrleansRuntime/Catalog/Catalog.cs

```diff
@@ -1328,6 +1329,7 @@ public void SiloStatusChangeNotification(SiloAddress updatedSilo, SiloStatus sta
                         try
                         {
                             var activationData = activation.Value;
+                            if (!activationData.IsUsingGrainDirectory) continue;
```

Comment:
That line was the only real fix. All the rest is refactoring, to decouple `IsStatelessWorker` from `IsUsingGrainDirectory`.